src-ui.v3/src/Main.hs

```diff
@@ -731,11 +736,72 @@ pkgTagList m = Map.fromListWith (List.++) $ do
+  | otherwise            = filter (JSS.isInfixOf (stripSearch sJ))pkgs
+
+splitInfixPkg :: JSS.JSString -> JSS.JSString -> (T.Text, T.Text, T.Text)
+splitInfixPkg stripSJ pkg = (frontT, midT, backT)
+  where
+    textS               = JSS.textFromJSString stripSJ
+    (frontT, reminderT) = T.breakOn textS (JSS.textFromJSString pkg)
+    (midT, backT)       = T.breakOnEnd textS reminderT
+
+calcMatchesNew :: [JSS.JSString] -> JSS.JSString -> Matches
+calcMatchesNew pkgs sJss = 
+  if JSS.length sJss < 3 
+  then matchesEmpty
+  else Matches { matchesInput = textS, matchesExact = exactMap, matchesInfix = othersMap}
+  where
+    textS                = JSS.textFromJSString sJss
+    stripSJ              = stripSearch sJss
+    (exactPkg, infixPkg) = List.partition (== stripSJ) pkgs
+    exactMap             = (Map.fromList . fmap (\p -> (JSS.textFromJSString p,()))) exactPkg
+    othersMap            = (Map.fromList . fmap (\p -> (JSS.textFromJSString p, splitInfixPkg stripSJ p))) (filterBySearchBox infixPkg sJss)
+
+searchBoxWidget :: forall t m. (SupportsServantReflex t m, MonadFix m, MonadIO m, MonadHold t m, PostBuild t m, DomBuilder t m, Adjustable t m, DomBuilderSpace m ~ GhcjsDomSpace)
+                => Dynamic t (Vector PkgN) 
+                -> m ()  
+searchBoxWidget dynPkgs0 = mdo
+  searchInputE <- elAttr "div" ("class" =: "item search right clearfix") $ do
+    divClass "text" $ text "Package Search"
+    sVal0 <- inputElement $ def & inputElementConfig_elementConfig . elementConfig_initialAttributes .~ fold ["class" =: "input-search","placeholder" =: "search..."]
+                                & inputElementConfig_setValue .~ clickPkgE
+    debounce 0.1 $ leftmost [clickPkgE, (_inputElement_input sVal0)]
+  let
+    dynPackagesJss = V.toList . fmap (JSS.textToJSString . pkgNToText) <$> dynPkgs0
+    matchesE = calcMatchesNew <$> current dynPackagesJss <@> (JSS.textToJSString <$> searchInputE)
+  matchesDyn <- holdDyn matchesEmpty matchesE
+  clickPkgE <- searchResultWidget matchesDyn
+  pure ()
+
+searchResultWidget :: forall t m. (MonadFix m, MonadHold t m, PostBuild t m, DomBuilder t m) 
+                   => Dynamic t Matches 
+                   -> m (Event t T.Text)
+searchResultWidget mDyn =
+  el "ul" $ do
+    exactE <- listViewWithKey (matchesExact <$> mDyn) $ \eId _ -> do
+                (e, _) <- element "li" def $ elAttr "a" ("href" =: ("#/package/" <> eId)) $ el "strong" $ text eId
+                pure $ domEvent Click e
+    otherE <- listViewWithKey (matchesInfix <$> mDyn) $ \pId txt -> do
+                (e, _) <- element "li" def $ elAttr "a" ("href" =: ("#/package/" <> pId)) $ do
+                            dynText . fmap (^. _1) $ txt
+                            el "strong" $ dynText . fmap (^. _2) $ txt
+                            dynText . fmap (^. _3) $ txt
+                pure $ domEvent Click e
+    pure $ "" <$ leftmost [exactE, otherE]
```

Comment:
This is about the point where I'd just make this an EventWriter to tidy it up. It's borderline though. :slightly_smiling_face: